src/coreclr/src/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -373,51 +481,90 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
             }
 
             // We've probed to find an empty spot, add to hash
-            if (!TryWriteValueToLocation(value, hashTableLocal, tableIndex))
+            if (!TryWriteSentinelToLocation(hashTableLocal, tableIndex))
             {
+                // After finding the empty spot, it was taken by some other thread
                 Interlocked.Decrement(ref _reserve);
                 return null;
             }
 
             // Now that we've written to the local array, find out if that array has been
             // replaced by expansion. If it has, we need to restart and write to the new array.
-            if (_newHashTable != hashTableLocal)
+            if (Interlocked.CompareExchange(ref _newHashTable, hashTableLocal, hashTableLocal) != hashTableLocal)
```

Comment:
This could be `if (_newHashTable != hashTableLocal)`  
You've just had a full fence before - when installing sentinel. `_newHashTable` will be fresh enough.